### PR TITLE
New strategy to generate roles

### PR DESCRIPTION
**Pull request by maidxn**
# Analyse the different between three strategies

- Strategy suggested by @nitro2 - Fixed ranged 
- Strategy suggested by @Stanley00  - Scored roles
- Stragegy suggested by @huybt22  - Handle conditions

Overview, all three methods can guarantee basic conditions like: 

- Villager party roles will have more players than Werewolf party and Third party.
- Always has seer and guard, make sure the game easier for villager.

But let's take a closer look in each method
## Fixed range

1. Because all roles have a fixed range, so it more likely that when the same players start a game, all roles will generate the same and it doesn't have flexibility. For example:

    - 1 lycan - 1 villager
    - 2 lycan 
    - 2 villager

    are the same meaning in Villager party, but when we use this method, we can't generate the 1 lycan - 1 villager scenario.

2. In one party, we can choose exactly the amount (range) for that party but not really check what roles are they, so it can cause like 2 superwolf in one game, and it is hard for the Villager party, makes the game unbalanced.

## Scored roles

In this method, all roles are represented by a score, the good party will have positive score, and the evil party (werewolf, fox, dracular,..) will have negative score. The aim of this method is to sum up all roles' scores to not greater than a specific value. 

It is hard to define what is the best value is the different between the good and bad party.

A plus point for this method is the board game Werewolf in real life is used this method to setup roles for players.

## Handle conditions

This method is checking if we add one role into our set roles, does it satisfy all the condition follow.

The down side of this method is when it comes with more new roles, we need to list all conditions and relative regard to that role, and handle all of that conditions.

## Personal conclusion
After analyzed, I recommend we use the second strategy - Scored roles to generate our roles in game. it has the flexibility, and it also can handle conditions based on the scored so we not really care about the conditions, we just need to make sure that we can define a good delta (diff between good party and evil party ) that can keeps the game balanced without worrying about to many other conditions.

Now I'm using the scores based on this site [The scores in Werewolf ultimate board game](https://ultimatewerewolfgames.tumblr.com/roles)

The map for role and score is here:
```
    role_dict = {
        "Seer": 7,
        "Witch": 6,
        "Guard": 3,
        "Hunter": 3,
        "Zombie": 3,
        "Chief": 2,
        "Villager": 1,
        "Tanner": 1,
        "Lycan": -1,
        "Fox": -1,
        "Betrayer": -2,
        "Cupid": -3,
        "Werewolf": -6,
        "Superwolf": -7,
    }
```

And because the fixed roles in our game is ['Werewolf', 'Seer', 'Guard', 'Villager'], so the scores in one simple game is = -6 + 7 + 3 + 1 = 5 (This is a basic scenario) and the max value is 7 (Seer role), so if sum of all roles in range(5, 7), that set of roles is acceptable.

This is the test code to generate roles from 5 to 50 players, each number of players I generated 10 times to check the role, your guys can check it here and please give me some advice to improve this better.

Test generated roles code [I am the code](https://tio.run/##rVXLbptAFN3zFVfZgDGJnFcjWXEXjdR2UXXTqpFqIURgMNPCDBqGJlaUb3fnzgwwJo7VRVlg@9zXuU83W1lydrnb0brhQoJIWc5rrxC8hoxXFckk5awFK77jHZNEeF5OCtgQmbQZFyQQvCJJTjMZKZARkUqSJwi2s6UH6hFEdoJB29Wj7lrEUHABAih7ZWYiZCXJficKp9zggev0kVdF0qRCbmEFa/@eCIKQH4H/rWuI6H98IFKkWyL8WJvJkorcsfueMhUbFe@6hub45SN/ssoZJpygp8h@H82V9SKCBWhFnYkih8lokoYjPrQYJCPlUbwfBuYrY7/Gdzxokcpx46RwyI9D8YA324tjmfX9NT0xpU8YeUxaicBmG9C87wLr6oSrpCoscatqUhGmxVpaE5kmeSpTJXj2v3Z1C1bTX05MX0wd6VM/BK@7SkyfPnWp0H36Qasq3QydVRWaslnB1VghmzkOX2AnOXDizQzlYUCR8mB7gsFPlnATjdA9lVmpsHcOprkp7NLBPutQE/Anrx8omYB3JSWFwi4crE9SwecObMZ2An7ZZilT2KkLqmmeQv1KIO7G0huAoEuqbwHibqrDkqHA1sU0MSeV7vjVWNCKtlhQ/AjEuPdjsXG8gXGJqNsU7/CUTYBTPXav7R5LqlYGZQOLGdxOrMMLZ0YGsvPp@Oni@jHMFT5OXnhhIpnLedaWXVFUxAmnpfqljyWSP3g53cmfjzzWy32y8V5m1uWtLbkqq0Xeww04WR0l9z@57S392v@m3aqirWyEYSIw1lubeDzG7F8u81t/HVOC6iQ9EAG8wNv8x1Idfb@trqOBvsGOjXNDnYvzHIaDjwjCUFN58TxT8XXs4T5QvQ8p25DgOoLrc3tfUfRrFJ0vZs4/S67tFzGEQAdUkrrRjTx@vgf9RlAmA7SaOaeyPUubhrB8IvB2u78)

Thank you so much for reading to the end of this longgggg note.

### CODE REVIEW

**Comment by nitro2** [comment]
Could you print your evaluate score in your test code?

File: game/generate_roles.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        "Villager": 1,
+        "Tanner": 1,
+        "Lycan": -1,
+        "Fox": -1,
+        "Betrayer": -2,
+        "Cupid": -3,
+        "Werewolf": -6,
+        "Superwolf": -7,
+    }
+    delta = 5
```
Comment (nitro2) [inline game/generate_roles.py:31]: Should we put these values to config file? cc @neihousaigaai

File: game/__init__.py [hunk 2]
```diff
@@ -179,21 +179,19 @@ def add_default_roles(self, role_json_in_string):
...
             yield from ('Werewolf' if i % 4 == 0 else 'Villager' for i in range(number - sum(cfg.values())))
 
         if self.runtime_roles:
             role_config = self.runtime_roles
         else:
-            role_config = utils.common.read_json_file("json/role_config.json")
+            role_config = generate_roles.generate_roles_new_strategy(ids)
 
         ids = list(ids)
         try:
```
Comment (Stanley00) [inline game/__init__.py:193]: We won't need `try...except` here anymore

File: game/generate_roles.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    num_of_players = num_of_players - len(fixed_roles)
+
+    while len(role_list) < num_of_players * 2:
+        role_list += ['Werewolf', 'Lycan'] + ['Villager'] * 2
+
+    random.shuffle(role_list)
+
+    scores = get_score(role_dict, fixed_roles + role_list[:num_of_players])
+
+    while scores < delta or scores > 7 :
```
Comment (Stanley00) [inline game/generate_roles.py:43]: We should check for `absolute` value of `scores`. An example is game having 10 werewolves and 1 villager

File: game/generate_roles.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        scores = get_score(role_dict, fixed_roles + role_list[:num_of_players])
```
Comment (Stanley00) [inline game/generate_roles.py:44]: We should calculate the score **after** `shuffle`

File: game/generate_roles.py (hunk 1 above)
Comment (neihousaigaai) [reply game/generate_roles.py]: Done moved to `role_generator_info.json` file.

File: game/__init__.py [hunk 5]
```diff
@@ -179,23 +179,18 @@ def add_default_roles(self, role_json_in_string):
             return "Invalid json format."
 
     def generate_roles(self, interface, ids, names_dict):
+
         def dict_to_list(cfg, number=0):
             yield from (name for name in cfg for _ in range(cfg[name]))
             yield from ('Werewolf' if i % 4 == 0 else 'Villager' for i in range(number - sum(cfg.values())))
```
Comment (nitro2) [inline game/__init__.py:185]: Could you help me move this to utils.py ? Same code in calc_points.py file

File: json/role_generator_info.json [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    "Chief": 2,
+    "Villager": 1,
+    "Tanner": 1,
+    "Lycan": -1,
+    "Fox": -1,
+    "Betrayer": -2,
+    "Cupid": -3,
+    "Werewolf": -6,
+    "Superwolf": -7
+  },
```
Comment (nitro2) [inline json/role_generator_info.json:17]: Please update the score reflect to https://docs.google.com/spreadsheets/d/10cbGmyFc1COK0eVkhC4FFXcRfnD7YvxRSzxA46NDIhU/edit#gid=4336794

File: game/__init__.py [hunk 7]
```diff
@@ -179,23 +179,14 @@ def add_default_roles(self, role_json_in_string):
...
     def generate_roles(self, interface, ids, names_dict):
-        def dict_to_list(cfg, number=0):
-            yield from (name for name in cfg for _ in range(cfg[name]))
-            yield from ('Werewolf' if i % 4 == 0 else 'Villager' for i in range(number - sum(cfg.values())))
 
         if self.runtime_roles:
             role_config = self.runtime_roles
         else:
-            role_config = utils.common.read_json_file("json/role_config.json")
+            role_config = generate_roles.generate_roles_new_strategy(ids)
```
Comment (nitro2) [inline game/__init__.py:186]: just a small change: Please comment the previous method to keep history reference:
```
            # Use fixed configuration file:
            # role_config = utils.common.read_json_file("json/role_config.json")
            # Use dynamic score system:
            role_config = generate_roles.generate_roles_new_strategy(ids)

```

Comment (nitro2) [reply game/__init__.py]: I updated

**Review by nitro2 — APPROVED** [review]

**Pull request merged**